Relax set_num_threads restriction in parallel native case

ilia-cher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #27925 Add CI builds for TBB and native parallel builds
* **#27947 Relax set_num_threads restriction in parallel native case**

Summary:
Don't throw exception if the requested size is the same as the currently
used one

Test Plan:
ATEN_THREADING=NATIVE python setup.py develop --cmake

Reviewers:liujiakai

Differential Revision: [D17919416](https://our.internmc.facebook.com/intern/diff/D17919416)

Commit:
Relax set_num_threads restriction in parallel native case

Summary:
Don't throw exception if the requested size is the same as the currently
used one

Test Plan:
ATEN_THREADING=NATIVE python setup.py develop --cmake

Reviewers:liujiakai

[ghstack-poisoned]

Commit:
Update on "Relax set_num_threads restriction in parallel native case"


Summary:
Don't throw exception if the requested size is the same as the currently
used one

Test Plan:
ATEN_THREADING=NATIVE python setup.py develop --cmake

Reviewers:liujiakai

[ghstack-poisoned]

dskhudia commented on aten/src/ATen/ParallelNative.cpp:
```diff
@@ -178,10 +178,19 @@ void set_num_threads(int nthreads) {
-      "Error: cannot set number of intraop threads "
-      "after parallel work has started or after set_num_threads call "
-      "when using native parallel backend");
+  if (!num_intraop_threads.compare_exchange_strong(no_value, nthreads)) {
```
So now calling, for example, set_num_threads(2) twice doesn't raise exception but set_num_threads(2) and then again set_num_threads(4) still raises exception with native thread pool. Correct?

ilia-cher replied:
yes, this is correct. The main reason for this is that we actually do have places that call set_num_threads(x) multiple times (where x is usually 1), and as of now, we may assume this to be a supported behavior. Changing number of threads though is not supported in native backend

Commit:
Update on "Relax set_num_threads restriction in parallel native case"


Summary:
Don't throw exception if the requested size is the same as the currently
used one

Test Plan:
ATEN_THREADING=NATIVE python setup.py develop --cmake

Reviewers:liujiakai

Differential Revision: [D17919416](https://our.internmc.facebook.com/intern/diff/D17919416)

[ghstack-poisoned]

ilia-cher commented:
using warning instead of exception

ilia-cher commented:
Regarding throwing exception vs. showing warning, it seems a bunch of tests are still broken, because of some use of set_num_threads, although real use cases should be fine afaik. We might try just showing a warning instead of throwing exception

Commit:
Update on "Relax set_num_threads restriction in parallel native case"


Summary:
Don't throw exception if the requested size is the same as the currently
used one

Test Plan:
ATEN_THREADING=NATIVE python setup.py develop --cmake

Reviewers:liujiakai

Differential Revision: [D17919416](https://our.internmc.facebook.com/intern/diff/D17919416)

[ghstack-poisoned]

Commit:
Update on "Relax set_num_threads restriction in parallel native case"


Summary:
Don't throw exception if the requested size is the same as the currently
used one

Test Plan:
ATEN_THREADING=NATIVE python setup.py develop --cmake

Reviewers:liujiakai

Differential Revision: [D17919416](https://our.internmc.facebook.com/intern/diff/D17919416)

[ghstack-poisoned]

Commit:
Update on "Relax set_num_threads restriction in parallel native case"


Summary:
Don't throw exception if the requested size is the same as the currently
used one

Test Plan:
ATEN_THREADING=NATIVE python setup.py develop --cmake

Reviewers:liujiakai

Differential Revision: [D17919416](https://our.internmc.facebook.com/intern/diff/D17919416)

[ghstack-poisoned]
